### PR TITLE
Add support for oci cli metrics

### LINKED ISSUE #262: Add support for OCI Client metrics
**What would you like to be added**:

Add support for OCI Client metrics

**Why is this needed**:

Customers can monitor calls to OCI API

### PR DESCRIPTION
**Pull request by shyamradhakrishnan**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #262

### CODE REVIEW

**Comment by shyamradhakrishnan** [comment]
unit test
```
?   	github.com/oracle/cluster-api-provider-oci	[no test files]
ok  	github.com/oracle/cluster-api-provider-oci/api/v1beta1	41.805s	coverage: 24.1% of statements
ok  	github.com/oracle/cluster-api-provider-oci/api/v1beta2	0.040s	coverage: 15.8% of statements
?   	github.com/oracle/cluster-api-provider-oci/cloud/config	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/metrics	[no test files]
ok  	github.com/oracle/cluster-api-provider-oci/cloud/ociutil	0.003s	coverage: 19.6% of statements
ok  	github.com/oracle/cluster-api-provider-oci/cloud/scope	241.272s	coverage: 76.3% of statements
?   	github.com/oracle/cluster-api-provider-oci/cloud/scope/mocks	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/base	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/base/mock_base	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/compute	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/compute/mock_compute	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/computemanagement	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/computemanagement/mock_computemanagement	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/containerengine	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/containerengine/mock_containerengine	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/identity	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/identity/mock_identity	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/loadbalancer	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/loadbalancer/mock_lb	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/networkloadbalancer	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/networkloadbalancer/mock_nlb	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/vcn	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/cloud/services/vcn/mock_vcn	[no test files]
ok  	github.com/oracle/cluster-api-provider-oci/cloud/util	0.331s	coverage: 76.1% of statements
ok  	github.com/oracle/cluster-api-provider-oci/controllers	26.269s	coverage: 67.5% of statements
ok  	github.com/oracle/cluster-api-provider-oci/exp/api/v1beta1	27.532s	coverage: 21.8% of statements
ok  	github.com/oracle/cluster-api-provider-oci/exp/api/v1beta2	0.026s	coverage: 10.9% of statements
ok  	github.com/oracle/cluster-api-provider-oci/exp/controllers	0.965s	coverage: 56.3% of statements
?   	github.com/oracle/cluster-api-provider-oci/feature	[no test files]
?   	github.com/oracle/cluster-api-provider-oci/version	[no test file
```

File: cloud/metrics/http_request_dispatcher_wrapper.go [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+import (
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/oracle/oci-go-sdk/v65/common"
+)
+
+// DispatcherWrapper is a wrapper around standard common.HTTPRequestDispatcher to handle
```
Comment (joekr) [inline cloud/metrics/http_request_dispatcher_wrapper.go:27]:
```suggestion
// HttpRequestDispatcherWrapper is a wrapper around standard common.HTTPRequestDispatcher to handle
```

File: cloud/scope/clients.go [hunk 2]
```diff
@@ -19,6 +19,7 @@ package scope
 import (
 	"crypto/tls"
 	"crypto/x509"
+	"github.com/oracle/cluster-api-provider-oci/cloud/metrics"
```
Comment (joekr) [inline cloud/scope/clients.go:22]: move this down?

File: cloud/metrics/http_request_dispatcher_wrapper.go [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+		}
+		resource := urlSplit[2]
+		IncRequestCounter(err, resource, req.Method, wrapper.region, resp)
+		ObserverRequestDuration(resource, req.Method, wrapper.region, time.Since(t))
+	}()
+	return resp, err
+}
+
+// NewHttpRequestDispatcherWrapper creates a new instance of HttpRequestDispatcherWrapper
+func NewHttpRequestDispatcherWrapper(wrapper common.HTTPRequestDispatcher, region string) HttpRequestDispatcherWrapper {
```
Comment (joekr) [inline cloud/metrics/http_request_dispatcher_wrapper.go:53]: Maybe rename `wrapper` to `dispatcher`. Isn't that parameter what is being wrapped?

**Review by joekr — CHANGES_REQUESTED** [review]
Some small nits

File: cloud/metrics/http_request_dispatcher_wrapper.go (hunk 1 above)
Comment (shyamradhakrishnan) [reply cloud/metrics/http_request_dispatcher_wrapper.go]: fixed

File: cloud/metrics/metrics.go [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+	ociRequestDurationSeconds = prometheus.NewHistogramVec(prometheus.HistogramOpts{
+		Subsystem: SubSystemOCI,
+		Name:      Duration,
+		Help:      "Duration/Latency of HTTP requests to OCI",
+	}, []string{Resource, Operation, Region})
+)
+
+// IncRequestCounter increments the request count metric for the given resource
+func IncRequestCounter(err error, resource string, operation string, region string, response *http.Response) {
+	// unknown errors from request dispatcher will have response code of 999
```
[the remark this answers is not in the document]
Comment (shyamradhakrishnan) [reply cloud/metrics/metrics.go]: fixed

File: cloud/scope/clients.go (hunk 2 above)
Comment (shyamradhakrishnan) [reply cloud/scope/clients.go]: fixed

File: cloud/metrics/http_request_dispatcher_wrapper.go (hunk 3 above)
Comment (shyamradhakrishnan) [reply cloud/metrics/http_request_dispatcher_wrapper.go]: fixed

**Review by joekr — APPROVED** [review]

**Pull request merged**